staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/typechecking.go

```diff
@@ -0,0 +1,406 @@
+	}
+
+	return allWarnings
+}
+
+// formatWarning converts the resulting issues and possible error during
+// type checking into a human-readable string
+func (c *TypeChecker) formatWarning(results []typeCheckingResult) string {
+	var sb strings.Builder
+	for _, result := range results {
+		if result.issues == nil && result.err == nil {
+			continue
+		}
+		if result.err != nil {
+			sb.WriteString(fmt.Sprintf("%v: type checking error: %v\n", result.gvk, result.err))
+		} else {
+			sb.WriteString(fmt.Sprintf("%v: %s\n", result.gvk, result.issues))
+		}
+	}
+	return sb.String()
+}
+
+func (c *TypeChecker) declType(gvk schema.GroupVersionKind) (*apiservercel.DeclType, error) {
+	if gvk.Empty() {
+		return nil, nil
+	}
+	s, err := c.schemaResolver.ResolveSchema(gvk)
+	if err != nil {
+		return nil, err
+	}
+	return common.SchemaDeclType(&openapi.Schema{Schema: s}, true), nil
+}
+
+func (c *TypeChecker) paramsType(policy *v1alpha1.ValidatingAdmissionPolicy) schema.GroupVersionKind {
+	if policy.Spec.ParamKind == nil {
+		return schema.GroupVersionKind{}
+	}
+	gv, err := schema.ParseGroupVersion(policy.Spec.ParamKind.APIVersion)
+	if err != nil {
+		return schema.GroupVersionKind{}
+	}
+	return gv.WithKind(policy.Spec.ParamKind.Kind)
+}
+
+func (c *TypeChecker) checkExpression(expression string, hasParams bool, types typeOverwrite) (*cel.Issues, error) {
+	env, err := buildEnv(hasParams, types)
+	if err != nil {
+		return nil, err
+	}
+
+	// We cannot reuse an AST that is parsed by another env, so reparse it here.
+	// Compile = Parse + Check, we especially want the results of Check.
+	//
+	// Paradoxically, we discard the type-checked result and let the admission
+	// controller use the dynamic typed program.
+	// This is a compromise that is defined in the KEP. We can revisit this
+	// decision and expect a change with limited size.
+	_, issues := env.Compile(expression)
+	return issues, nil
+}
+
+// typesToCheck extracts a list of GVKs that needs type checking from the policy
+// the result is sorted in the order of Group, Version, and Kind
+func (c *TypeChecker) typesToCheck(p *v1alpha1.ValidatingAdmissionPolicy) []schema.GroupVersionKind {
+	gvks := sets.New[schema.GroupVersionKind]()
+	if p.Spec.MatchConstraints == nil || len(p.Spec.MatchConstraints.ResourceRules) == 0 {
+		return nil
+	}
+
+	for _, rule := range p.Spec.MatchConstraints.ResourceRules {
+		groups := extractGroups(&rule.Rule)
+		if len(groups) == 0 {
+			continue
+		}
+		versions := extractVersions(&rule.Rule)
+		if len(versions) == 0 {
+			continue
+		}
+		resources := extractResources(&rule.Rule)
+		if len(resources) == 0 {
+			continue
+		}
+		for _, group := range groups {
+			for _, version := range versions {
+				for _, resource := range resources {
+					gvr := schema.GroupVersionResource{
+						Group:    group,
+						Version:  version,
+						Resource: resource,
+					}
+					resolved, err := c.restMapper.KindsFor(gvr)
```

Comment:
do we know how much this will get called and how expensive it is?

not sure which restmapper implementation is underlying this, but restmapper lookup can be really expensive when it hits discovery, and lookup failures trigger clearing the cache and re-hitting discovery in some implementations

calling this inside a quadruple-nested loop called from a controller sync makes me very anxious

Reply:
I agree we should do something about excessively inclusive type matches.
However, even if the controller caches result of `KindsFor`, since the types in a cluster is unbounded, the user can still evade the cache by request different types each time.

Would you suggest limit the maximum types to check in this case? I can update the KEP to include this limitation in next release.

Reply:
I am starting to work on setting a limit of 10 types to check. Please let me know if that decision does not fully address the problem. Thank you.

Reply:
Is the limit documented in the API documentation? If the limit is exceeded, the 1st 10 types are still checked? I wonder if this will create confusion when the 1st 10 types have no errors but, say, the 15th type does. Will this report no errors in that case or will it report in some way that type checking was only performed for some of the types?

Reply:
It will report no errors. Do we expect a policy to have so many manually specified resource types? I can only think of the use case where a policy can catch the same type of different versions. Otherwise, the user should put wildcards. Types that match wildcards do not get checked, only the ones that are exactly specified in the constraints are.

Reply:
I expect >10 to be very rare. I agree that wildcards should not be type checked.

Reply:
If that's the case, let's document the limitation, and proceed with limit the check to the first 10 types.

Reply:
bounding the number of times we call this sounds good. 

> not sure which restmapper implementation is underlying this, but restmapper lookup can be really expensive when it hits discovery, and lookup failures trigger clearing the cache and re-hitting discovery in some implementations

This bit is still relevant, even if we only call this once per policy we process. What happens if the gvr we ask about is an unknown one, and we perpetually ask about it... will that trigger constant full rediscovery by the restmapper?

Reply:
For now the answer is never. Now that CRD support is disabled, in all code paths I can trace, the RestMapper instances are all constructed from registering each known types. So I assume there is no roundtrip anymore.